contracts/marketpools.js

```diff
@@ -0,0 +1,348 @@
+
+  const tokenInAdjusted = api.BigNumber(getAmountIn(tokenOut, liquidityIn, liquidityOut));
+  if (!tokenInAdjusted.isFinite()) return;
+
+  const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: symbolIn });
+  const senderBaseFunded = api.BigNumber(senderBase.balance).gte(tokenInAdjusted);
+  const tokenPairDelta = tokenSymbol === baseSymbol ? [api.BigNumber(tokenOut).times(-1), tokenInAdjusted] : [tokenInAdjusted, api.BigNumber(tokenOut).times(-1)];
+  if (!api.assert(senderBaseFunded, 'insufficient input balance')
+    || !validateSwap(pool, tokenPairDelta[0], tokenPairDelta[1])) return;
+
+  await api.executeSmartContract('tokens', 'transferToContract', { symbol: symbolIn, quantity: tokenInAdjusted.toFixed(pool.precision), to: 'marketpools' });
+  await api.transferTokens(api.sender, symbolOut, tokenOut, 'user');
+  updatePoolStats(pool, tokenPairDelta[0], tokenPairDelta[1]);
+  api.emit('swapTokensForExactTokens', { memo: `Swap ${symbolIn} for ${symbolOut}` });
+};
+
+actions.swapExactTokensForTokens = async (payload) => {
+  const {
+    tokenPair,
+    tokenSymbol,
+    tokenIn,
+    isSignedWithActiveKey,
+  } = payload;
+
+  if (!api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
+    || !api.assert(typeof (tokenSymbol) === 'string', 'invalid token')
+    || !api.assert(tokenIn && api.BigNumber(tokenIn).gt(0), 'insufficient tokenIn')
+    || !await validateTokenPair(tokenPair)) {
+    return;
+  }
+
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  let liquidityIn;
+  let liquidityOut;
+  let symbolIn;
+  let symbolOut;
+  if (tokenSymbol === baseSymbol) {
+    liquidityIn = pool.baseQuantity;
+    liquidityOut = pool.quoteQuantity;
+    symbolIn = baseSymbol;
+    symbolOut = quoteSymbol;
+  } else {
+    liquidityIn = pool.quoteQuantity;
+    liquidityOut = pool.baseQuantity;
+    symbolIn = quoteSymbol;
+    symbolOut = baseSymbol;
+  }
+
+  if (!api.assert(pool, 'no existing pool for tokenPair')) return;
+
+  const tokenOutAdjusted = api.BigNumber(getAmountOut(tokenIn, liquidityIn, liquidityOut));
+  if (!tokenOutAdjusted.isFinite()) return;
+
+  const senderBase = await api.db.findOneInTable('tokens', 'balances', { account: api.sender, symbol: symbolIn });
+  const senderBaseFunded = api.BigNumber(senderBase.balance).gte(tokenIn);
+  const tokenPairDelta = tokenSymbol === baseSymbol ? [tokenIn, api.BigNumber(tokenOutAdjusted).times(-1)] : [api.BigNumber(tokenOutAdjusted).times(-1), tokenIn];
+  if (!api.assert(senderBaseFunded, 'insufficient input balance')
+    || !validateSwap(pool, tokenPairDelta[0], tokenPairDelta[1])) return;
+
+  await api.executeSmartContract('tokens', 'transferToContract', { symbol: symbolIn, quantity: tokenIn, to: 'marketpools' });
+  await api.transferTokens(api.sender, symbolOut, tokenOutAdjusted.toFixed(pool.precision), 'user');
```

Comment:
You should put in a sanity check that the transferToContract action succeeded, before transferring tokens to api.sender.  Also, I question the rounding to pool.precision.  Shouldn't that be to the out token's precision?  If pool.precision > out token precision, this could fail in the tokens contract with a "symbol precision mismatch" error.

Reply:
My thinking was setting pool precision to the smallest precision of either token in the pair and then using that throughout. It seemed simpler at time of writing to avoid mismatches but maybe there are some cases I didn't consider.

Reply:
Okay, I forgot that's how the pool precision works.  It should be safe then, as rounding to the pool precision would stay within the precision limits of both tokens.

Reply:
Both transfers updated to use the corresponding token's precision.